### PR TITLE
An assortment of fixes and improvements

### PR DESCRIPTION
**Pull request by elprans**
Changes:

### Add `edbdocker_check_instance` callback

The `edbdocker_check_instance` is intended to be overloaded in a derived 
image to check the validity of an existing database instance, e.g. to check
catalog version compatibility.

### Make `edbdocker_run_temp_server()` non-fatal by itself

Currently, `edbdocker_run_temp_server()` calls `die` on all errors. To
allow uses where server failure is expected, delegate the decision to abort
the whole thing to the abort callback.

### Fix entrypoint sourcing

Two fixes:
- `$FUNCNAME` is only defined inside a function, so make `_is_sourced`
- Use the topmost `$BASH_SOURCE` when restarting to drop privileges

### Apply migrations during bootstrap instead of restarting server

Remove the need to restart the server to apply the migrations when 
bootstrapping a new instance.

### CODE REVIEW

**Review by tailhook — COMMENTED** [review]
Well, I think this is getting to complex and to hard to reason about. The following would add to issue even more:
> Use the topmost $BASH_SOURCE when restarting to drop privileges

I think we should move functions out of `docker-entrypoint.sh`, remove the need to source entry point itself, and put clearly what's going on into an entrypoint file, along the lines of:
```
source functions.sh
if is_server_cmd "$@"; then
    if bootstrap_needed; then
        start_bootstrap
        setup_roles
        run_hooks
        run_migrations
        end_bootstrap
    fi
    start_server
else
    exec "$@"
fi
```
So that it's clear what's going on when you read `entrypoint.sh`. And you can copy it and add any extra things instead of defining hooks.

**Comment by elprans** [comment]
> I think we should move functions

Fair enough.

**Comment by elprans** [comment]
OK, I reorganized the entrypoint to source functions and dropped the need to restart the entire script.  Privilege drop is now applied granularly on server/CLI invocations.

File: docker-entrypoint-funcs.sh [hunk 1]
```diff
@@ -0,0 +1,662 @@
...
+  fi
+
+  edbdocker_die "$1"
+}
+
+
+# Runs on an existing (not freshly bootstrapped) instance, intended
+# to be overloaded in derived images for extra checks prior to server
+# startup.
+edbdocker_check_instance() {
```
Comment (tailhook) [inline docker-entrypoint-funcs.sh:433]: Is unused any more.

**Review by tailhook — APPROVED** [review]
Looks great!

I'm slightly concerned that migrations require a separate server run on each invocation (which is slow). For most of my use cases I would probably be okay with migrations running in background. But well, it's probably easy enough to customize, and is better to be more strict by default.

**Pull request merged**